# Advent of Code 2025 Day 4: Printing Department

## Import libraries

## Part 1

### Description Part 1

You ride the escalator down to the printing department. They're clearly getting ready for Christmas; they have lots of large rolls of paper everywhere, and there's even a massive printer in the corner (to handle the really big print jobs).

Decorating here will be easy: they can make their own decorations. What you really need is a way to get further into the North Pole base while the elevators are offline.

"Actually, maybe we can help with that," one of the Elves replies when you ask for help. "We're pretty sure there's a cafeteria on the other side of the back wall. If we could break through the wall, you'd be able to keep moving. It's too bad all of our forklifts are so busy moving those big rolls of paper around."

If you can optimize the work the forklifts are doing, maybe they would have time to spare to break through the wall.

The rolls of paper (@) are arranged on a large grid; the Elves even have a helpful diagram (your puzzle input) indicating where everything is located.

For example:

..@@.@@@@.
@@@.@.@.@@
@@@@@.@.@@
@.@@@@..@.
@@.@@@@.@@
.@@@@@@@.@
.@.@.@.@@@
@.@@@.@@@@
.@@@@@@@@.
@.@.@@@.@.
The forklifts can only access a roll of paper if there are fewer than four rolls of paper in the eight adjacent positions. If you can figure out which rolls of paper the forklifts can access, they'll spend less time looking and more time breaking down the wall to the cafeteria.

In this example, there are 13 rolls of paper that can be accessed by a forklift (marked with x):

..xx.xx@x.
x@@.@.@.@@
@@@@@.x.@@
@.@@@@..@.
x@.@@@@.@x
.@@@@@@@.@
.@.@.@.@@@
x.@@@.@@@@
.@@@@@@@@.
x.x.@@@.x.
Consider your complete diagram of the paper roll locations. How many rolls of paper can be accessed by a forklift?

### Import data

In [1]:
with open('day_4_input.txt') as input:
    rolls = [line.strip() for line in input.read().split('\n') if line.strip()]

In [2]:
rolls[:10]

['@@@.@@@@@@@@@@@@.@.@..@@@@@@@@..@.@.@.@.@.....@@@@.@@@@..@@@@@.@@@.@@.@@@@.@@..@@@.@@@@.@.@@@@@.@@@@@@@@.@...@@@.@.@@@....@@.@@.@.@@@@.@@@.',
 '.@.@@....@.@.@..@..@@@.@.@...@......@@@.@@@.@.@..@@@@@@....@@@.@@@@@@..@.@@...@@..@.@@@..@@.@@@.@@@@@@.@@.@@@@.@@@.@@@.@.@@@@..@.@..@..@@.@',
 '@@@.@@@@..@.@@...@@@@@.@.@..@@@@..@@.@@@@@@.@..@@@@@@@@@.@.@.@@..@.@.@@@@.@@@..@@@@@@@@@@@@@...@@.@...@@@@.@@.@@.@@.@@@@..@..@...@@.@@@.@.@',
 '@@.@@@.@@.@..@.@@@.....@@@@@@.@@@@@..@@@@@@.@@@..@.@.@@@.@.@@.@..@.@.....@....@..@@@.@.@@.@@.@@.@@@..@.@.@@@..@@.@@..@@@@@@.@@@@@.@@@@@@@.@',
 '@@@@@..@@@@@@@...@..@@@.@.@@@..@@.@..@.@@@.@..@@..@.@@@.@@@@.@..@@@@@@@@@@@@@@@@@.@.@@@....@.@....@@@@@..@@.@@.@@@@..@@@.@@@.@@@.@@@..@.@@.',
 '.@.@..@.@@@@@.@@@@@@@@@@@@@@@@@@@@.@@..@@..@@@@@..@@@@.@@...@@@.@@..@@.@..@.@.@@.@@@@@@@@@@@.@@..@@@@..@@..@@.@.....@@@..@@.@@@@@.@.@..@@@@',
 '.@@@@.@.@@@@..@@@@@@@@@@@@.@@@@@@@@@..@..@@@@@@@@@@@..@.@..@@.....@@@.@@@@.@@..@@.@@@.@..@.@@..@@@.@@@@.@@.@..@@...@.@@@..@.@@@@@@@@.@

In [3]:
# Create test set (Answer for part 1 is 13)
test_list = ['..@@.@@@@.', '@@@.@.@.@@', '@@@@@.@.@@', '@.@@@@..@.', '@@.@@@@.@@', '.@@@@@@@.@', '.@.@.@.@@@', '@.@@@.@@@@', '.@@@@@@@@.', '@.@.@@@.@.']

### Write function to ID & count all rolls of paper with fewer than 4 surrounding rolls of paper

In [4]:
# Using while loop, write fn to get:
    # 3 characters above current character
    # 2 characters on either side of current character
    # 3 characters below current character
# Add each to an object, count the number of @ symbols in object
# If number of @ symbols < 4, + 1 to total_sum

def surrounding_rolls(input):

    # Create object to store the sum of all rolls with 4+ surrounding rolls
    total_sum = 0
    # Create an index to iterate through the lines
    input_index = 0
    # Iterate through the list of lines of rolls
    while input_index < len(input):
        # Only define previous line if the current line is not the first line in the list
        if input_index > 0:
            # Define the previous line
            prev_line = input[input_index - 1]
        # Define the current line
        line = input[input_index]
        # Only define next line if the current line is not the final line in the list
        if input_index < len(input) - 1:
            # Define the next line
            next_line = input[input_index + 1]
        # Create an index to iterate through the rolls in current line
        line_index = 0
        # Iterate through the characters in line
        while line_index < len(line):
            # Do not check for surrounding rolls if the current character is a '.'
            if line[line_index] != '.':
                # Define a string object to put the surrounding roll characters into
                surrounding_rolls = ''
                # For any lines that are not the first or final lines in the list
                if line_index > 0 and line_index < len(line) - 1:
                    # If the current line is not the first line (no lines before first line)
                    if input_index > 0:
                        # Fetch the previous line characters
                        prev_line_rolls = str(prev_line[line_index - 1]) + str(prev_line[line_index]) + str(prev_line[line_index + 1])
                        # Add to the object containing all surrounding rolls
                        surrounding_rolls += str(prev_line_rolls)
                    # Fetch the current line characters (same for all lines)
                    current_line_rolls = str(line[line_index - 1]) + str(line[line_index + 1])
                    # Add to the object containing all surrounding rolls
                    surrounding_rolls += str(current_line_rolls)
                    # If the current line is not the final line (no lines after final line)
                    if input_index < len(input) - 1:
                        # Fetch the next line characters
                        next_line_rolls = str(next_line[line_index - 1]) + str(next_line[line_index]) + str(next_line[line_index + 1])
                        # Add to the object containing all surrounding rolls
                        surrounding_rolls += str(next_line_rolls)
                # Special case for first column characters- no characters to left
                elif line_index == 0:
                    # If the current line is not the first line (no lines before first line)
                    if input_index > 0:
                        # Fetch the previous line characters
                        prev_line_rolls = str(prev_line[line_index]) + str(prev_line[line_index + 1])
                        # Add to the object containing all surrounding rolls
                        surrounding_rolls += str(prev_line_rolls)
                    # Fetch the current line characters
                    current_line_rolls = str(line[line_index + 1])
                    # Add to the object containing all surrounding rolls
                    surrounding_rolls += str(current_line_rolls)
                    # If the current line is not the final line (no lines after final line)
                    if input_index < len(input) - 1:
                        # Fetch the next line characters
                        next_line_rolls = str(next_line[line_index]) + str(next_line[line_index + 1])
                        # Add to the object containing all surrounding rolls
                        surrounding_rolls += str(next_line_rolls)
                # Special case for last column characters- no characters to right
                elif line_index == len(line) - 1:
                    # If the current line is not the first line (no lines before first line)
                    if input_index > 0:
                        # Fetch the previous line characters
                        prev_line_rolls = str(prev_line[line_index - 1]) + str(prev_line[line_index])
                        # Add to the object containing all surrounding rolls
                        surrounding_rolls += str(prev_line_rolls)
                    # Fetch the current line characters
                    current_line_rolls = str(line[line_index - 1])
                    # Add to the object containing all surrounding rolls
                    surrounding_rolls += str(current_line_rolls)
                    # If the current line is not the final line (no lines after final line)
                    if input_index < len(input) - 1:
                        # Fetch the next line characters
                        next_line_rolls = str(next_line[line_index - 1]) + str(next_line[line_index])
                        # Add to the object containing all surrounding rolls
                        surrounding_rolls += str(next_line_rolls)
                # Remove all '.' from string to count number of @ symbols
                surrounding_rolls = surrounding_rolls.replace('.', '')
                # Count the number of @ symbols in string, if > 4 then add 1 to total
                if len(surrounding_rolls) < 4:
                    total_sum += 1
            # Proceed to the next character
            line_index += 1
        # Proceed to the next line
        input_index +=1

    return total_sum

In [5]:
test_list

['..@@.@@@@.',
 '@@@.@.@.@@',
 '@@@@@.@.@@',
 '@.@@@@..@.',
 '@@.@@@@.@@',
 '.@@@@@@@.@',
 '.@.@.@.@@@',
 '@.@@@.@@@@',
 '.@@@@@@@@.',
 '@.@.@@@.@.']

In [6]:
surrounding_rolls(rolls)

1587

## Solution: The number of rolls which can be accessed is 1,587

## Part 2

### Description Part 2

Now, the Elves just need help accessing as much of the paper as they can.

Once a roll of paper can be accessed by a forklift, it can be removed. Once a roll of paper is removed, the forklifts might be able to access more rolls of paper, which they might also be able to remove. How many total rolls of paper could the Elves remove if they keep repeating this process?

Starting with the same example as above, here is one way you could remove as many rolls of paper as possible, using highlighted @ to indicate that a roll of paper is about to be removed, and using x to indicate that a roll of paper was just removed:

Initial state:
..@@.@@@@.
@@@.@.@.@@
@@@@@.@.@@
@.@@@@..@.
@@.@@@@.@@
.@@@@@@@.@
.@.@.@.@@@
@.@@@.@@@@
.@@@@@@@@.
@.@.@@@.@.

Remove 13 rolls of paper:
..xx.xx@x.
x@@.@.@.@@
@@@@@.x.@@
@.@@@@..@.
x@.@@@@.@x
.@@@@@@@.@
.@.@.@.@@@
x.@@@.@@@@
.@@@@@@@@.
x.x.@@@.x.

Remove 12 rolls of paper:
.......x..
.@@.x.x.@x
x@@@@...@@
x.@@@@..x.
.@.@@@@.x.
.x@@@@@@.x
.x.@.@.@@@
..@@@.@@@@
.x@@@@@@@.
....@@@...

Remove 7 rolls of paper:
..........
.x@.....x.
.@@@@...xx
..@@@@....
.x.@@@@...
..@@@@@@..
...@.@.@@x
..@@@.@@@@
..x@@@@@@.
....@@@...

Remove 5 rolls of paper:
..........
..x.......
.x@@@.....
..@@@@....
...@@@@...
..x@@@@@..
...@.@.@@.
..x@@.@@@x
...@@@@@@.
....@@@...

Remove 2 rolls of paper:
..........
..........
..x@@.....
..@@@@....
...@@@@...
...@@@@@..
...@.@.@@.
...@@.@@@.
...@@@@@x.
....@@@...

Remove 1 roll of paper:
..........
..........
...@@.....
..x@@@....
...@@@@...
...@@@@@..
...@.@.@@.
...@@.@@@.
...@@@@@..
....@@@...

Remove 1 roll of paper:
..........
..........
...x@.....
...@@@....
...@@@@...
...@@@@@..
...@.@.@@.
...@@.@@@.
...@@@@@..
....@@@...

Remove 1 roll of paper:
..........
..........
....x.....
...@@@....
...@@@@...
...@@@@@..
...@.@.@@.
...@@.@@@.
...@@@@@..
....@@@...

Remove 1 roll of paper:
..........
..........
..........
...x@@....
...@@@@...
...@@@@@..
...@.@.@@.
...@@.@@@.
...@@@@@..
....@@@...
Stop once no more rolls of paper are accessible by a forklift. In this example, a total of 43 rolls of paper can be removed.

Start with your original diagram. How many rolls of paper in total can be removed by the Elves and their forklifts?

### Write function to ID accessible rolls & create new list where surrounding rolls are removed

In [7]:
# Same funcion as above, but if < 4 surrounding rolls, then change the current roll to '.'

def surrounding_rolls_remove(input, total_sum):
    
    # Create a new list to store the new characters
    new_list = []
    # Create an index to iterate through the lines
    input_index = 0
    # Iterate through the list of lines of rolls
    while input_index < len(input):
        # Only define previous line if the current line is not the first line in the list
        if input_index > 0:
            # Define the previous line
            prev_line = input[input_index - 1]
        # Define the current line
        line = input[input_index]
        # Only define next line if the current line is not the final line in the list
        if input_index < len(input) - 1:
            # Define the next line
            next_line = input[input_index + 1]
        # Create an index to iterate through the rolls in current line
        line_index = 0
        # Create a new string to store the new characters
        new_string = ''
        # Iterate through the characters in line
        while line_index < len(line):
            # If the current character is a '.', add it to the new string as is
            if line[line_index] == '.':
                new_string += line[line_index]
            # Do not check for surrounding rolls if the current character is a '.'
            elif line[line_index] != '.':
                # Define a string object to put the surrounding roll characters into
                surrounding_rolls = ''
                # For any lines that are not the first or final lines in the list
                if line_index > 0 and line_index < len(line) - 1:
                    # If the current line is not the first line (no lines before first line)
                    if input_index > 0:
                        # Fetch the previous line characters
                        prev_line_rolls = str(prev_line[line_index - 1]) + str(prev_line[line_index]) + str(prev_line[line_index + 1])
                        # Add to the object containing all surrounding rolls
                        surrounding_rolls += str(prev_line_rolls)
                    # Fetch the current line characters (same for all lines)
                    current_line_rolls = str(line[line_index - 1]) + str(line[line_index + 1])
                    # Add to the object containing all surrounding rolls
                    surrounding_rolls += str(current_line_rolls)
                    # If the current line is not the final line (no lines after final line)
                    if input_index < len(input) - 1:
                        # Fetch the next line characters
                        next_line_rolls = str(next_line[line_index - 1]) + str(next_line[line_index]) + str(next_line[line_index + 1])
                        # Add to the object containing all surrounding rolls
                        surrounding_rolls += str(next_line_rolls)
                # Special case for first column characters- no characters to left
                elif line_index == 0:
                    # If the current line is not the first line (no lines before first line)
                    if input_index > 0:
                        # Fetch the previous line characters
                        prev_line_rolls = str(prev_line[line_index]) + str(prev_line[line_index + 1])
                        # Add to the object containing all surrounding rolls
                        surrounding_rolls += str(prev_line_rolls)
                    # Fetch the current line characters
                    current_line_rolls = str(line[line_index + 1])
                    # Add to the object containing all surrounding rolls
                    surrounding_rolls += str(current_line_rolls)
                    # If the current line is not the final line (no lines after final line)
                    if input_index < len(input) - 1:
                        # Fetch the next line characters
                        next_line_rolls = str(next_line[line_index]) + str(next_line[line_index + 1])
                        # Add to the object containing all surrounding rolls
                        surrounding_rolls += str(next_line_rolls)
                # Special case for last column characters- no characters to right
                elif line_index == len(line) - 1:
                    # If the current line is not the first line (no lines before first line)
                    if input_index > 0:
                        # Fetch the previous line characters
                        prev_line_rolls = str(prev_line[line_index - 1]) + str(prev_line[line_index])
                        # Add to the object containing all surrounding rolls
                        surrounding_rolls += str(prev_line_rolls)
                    # Fetch the current line characters
                    current_line_rolls = str(line[line_index - 1])
                    # Add to the object containing all surrounding rolls
                    surrounding_rolls += str(current_line_rolls)
                    # If the current line is not the final line (no lines after final line)
                    if input_index < len(input) - 1:
                        # Fetch the next line characters
                        next_line_rolls = str(next_line[line_index - 1]) + str(next_line[line_index])
                        # Add to the object containing all surrounding rolls
                        surrounding_rolls += str(next_line_rolls)
                # Remove all '.' from string to count number of @ symbols
                surrounding_rolls = surrounding_rolls.replace('.', '')
                # Count the number of @ symbols in string, if < 4 then change the character to a '.'
                if len(surrounding_rolls) < 4:
                    new_string += '.'
                    # Add to total sum of removed rolls
                    total_sum += 1
                # Count the number of @ symbols in string, if > 4 then keep the character as an '@'
                elif len(surrounding_rolls) >= 4:
                    new_string += '@'
            # Proceed to the next character
            line_index += 1
        # Add to the new list of the updated strings
        new_list += [new_string]
        # Proceed to the next line
        input_index +=1

    return new_list, total_sum

In [8]:
# Use while loop to iterate function over subsequent outputs
# Thank you, @DavidVogelxyz- taught me how to do this part

# Set the sum to one above new sum- over each iteration, it will stay higher as long as something changes in the previous output
# Once new_sum catches up to sum, the loop will stop
sum = 0
new_sum = -1

# Iterate through each subsequent function output
while sum != new_sum:
    # Change new_sum to sum to step up (tring to make new_sum catch up to stop loop)
    new_sum = sum
    # Run the function
    rolls, sum = surrounding_rolls_remove(rolls, sum)

In [9]:
print(test_list, sum)

['..@@.@@@@.', '@@@.@.@.@@', '@@@@@.@.@@', '@.@@@@..@.', '@@.@@@@.@@', '.@@@@@@@.@', '.@.@.@.@@@', '@.@@@.@@@@', '.@@@@@@@@.', '@.@.@@@.@.'] 8946


## Solution: The total number of rolls that can be removed is 8,946.